sklearn/decomposition/tests/test_truncated_svd.py

```diff
@@ -7,227 +7,178 @@
 
 from sklearn.decomposition import TruncatedSVD, PCA
 from sklearn.utils import check_random_state
-from sklearn.utils.testing import (assert_array_almost_equal, assert_equal,
-                                   assert_raises, assert_greater,
-                                   assert_array_less, assert_allclose)
+from sklearn.utils.testing import assert_array_less, assert_allclose
 
+SVD_SOLVERS = ['arpack', 'randomized']
 
-# Make an X that looks somewhat like a small tf-idf matrix.
-# XXX newer versions of SciPy >0.16 have scipy.sparse.rand for this.
-shape = 60, 55
-n_samples, n_features = shape
-rng = check_random_state(42)
-X = rng.randint(-100, 20, np.product(shape)).reshape(shape)
-X = sp.csr_matrix(np.maximum(X, 0), dtype=np.float64)
-X.data[:] = 1 + np.log(X.data)
-Xdense = X.A
 
+@pytest.fixture(scope='module')
+def X_sparse():
+    # Make an X that looks somewhat like a small tf-idf matrix.
+    rng = check_random_state(42)
+    X = sp.random(60, 55, density=0.2, format="csr", random_state=rng)
+    X.data[:] = 1 + np.log(X.data)
+    return X
 
-def test_algorithms():
+
+@pytest.mark.parametrize("solver", ['randomized'])
+@pytest.mark.parametrize('kind', ('dense', 'sparse'))
+def test_solvers(X_sparse, solver, kind):
+    X = X_sparse if kind == 'sparse' else X_sparse.toarray()
     svd_a = TruncatedSVD(30, algorithm="arpack")
-    svd_r = TruncatedSVD(30, algorithm="randomized", random_state=42)
+    svd = TruncatedSVD(30, algorithm=solver, random_state=42)
 
     Xa = svd_a.fit_transform(X)[:, :6]
-    Xr = svd_r.fit_transform(X)[:, :6]
-    assert_array_almost_equal(Xa, Xr, decimal=5)
+    Xr = svd.fit_transform(X)[:, :6]
+    assert_allclose(Xa, Xr, rtol=2e-3)
 
     comp_a = np.abs(svd_a.components_)
-    comp_r = np.abs(svd_r.components_)
+    comp = np.abs(svd.components_)
     # All elements are equal, but some elements are more equal than others.
-    assert_array_almost_equal(comp_a[:9], comp_r[:9])
-    assert_array_almost_equal(comp_a[9:], comp_r[9:], decimal=2)
+    assert_allclose(comp_a[:9], comp[:9], rtol=1e-3)
+    assert_allclose(comp_a[9:], comp[9:], rtol=2e-1, atol=1e-2)
```

Comment:
Is `rtol` needed here?

Reply:
Yes, later coefficients need higher tolerance, also `decimal` in `assert_array_almost_equal` account for both rtol and atol I think?

Reply:
In the check for later coefficients:

```
assert_allclose(comp_a[9:], comp[9:], rtol=2e-1, atol=1e-2)
```

with `atol=1e-2` it feels like `rtol` can even be set to zero.

BTW I am fine either way, this is kind of nitpicky. 😅

Reply:
On a second look, you are right -- removed it.